src/libultra/libc/xprintf.c

```diff
@@ -131,85 +150,88 @@ void _Putfld(_Pft* px, va_list* pap, u8 code, u8* ac) {
                 ac[px->n0++] = ' ';
             }
 
-            px->s = (char*)&ac[px->n0];
+            px->s = &ac[px->n0];
 
             _Litob(px, code);
             break;
+
         case 'x':
         case 'X':
         case 'u':
         case 'o':
             if (px->qual == 'l') {
-                px->v.ll = va_arg(*pap, s32);
+                px->v.ll = va_arg(*pap, long);
             } else if (px->qual == 'L') {
-                px->v.ll = va_arg(*pap, s64);
+                px->v.ll = va_arg(*pap, long long);
             } else {
-                px->v.ll = va_arg(*pap, s32);
+                px->v.ll = va_arg(*pap, int);
             }
 
             if (px->qual == 'h') {
-                px->v.ll = (u16)px->v.ll;
-            } else if (px->qual == 0) {
-                px->v.ll = (u32)px->v.ll;
+                px->v.ll = (unsigned short)px->v.ll;
+            } else if (px->qual == '\0') {
+                px->v.ll = (unsigned int)px->v.ll;
             }
 
             if (px->flags & FLAGS_HASH) {
                 ac[px->n0++] = '0';
                 if (code == 'x' || code == 'X') {
-
                     ac[px->n0++] = code;
                 }
             }
-            px->s = (char*)&ac[px->n0];
+            px->s = &ac[px->n0];
             _Litob(px, code);
             break;
+
         case 'e':
         case 'f':
         case 'g':
         case 'E':
         case 'G':
-            px->v.ld = px->qual == 'L' ? va_arg(*pap, f64) : va_arg(*pap, f64);
+            px->v.ld = px->qual == 'L' ? va_arg(*pap, LONG_DOUBLE_TYPE) : va_arg(*pap, double);
 
-            if (*(u16*)&px->v.ll & 0x8000) {
+            if (LDSIGN(px->v.ld)) {
```

Comment:
why make a macro for this? and is there a reason to not do `ld < 0` (besides, for us, matching)

this with the macro `isdigit` (not `IS_DIGIT`) makes me wonder if you yoinked this code from somewhere or are looking at a reference?

Reply:
I guess they wanted to avoid float operations, possibly for NaN/infinity handling. I tried to sync a bit more with ultralib's xprintf hence the macros